cdqa/reader/bertqa_sklearn.py

```diff
@@ -1257,3 +1261,93 @@ def predict(self, X, return_logit=False):
+            if len(all_results) % 1000 == 0 and self.verbose_logging:
+                logger.info("Processing example: %d" % (len(all_results)))
+            input_ids = input_ids.to(self.device)
+            input_mask = input_mask.to(self.device)
+            segment_ids = segment_ids.to(self.device)
+            with torch.no_grad():
+                batch_start_logits, batch_end_logits = self.model(
+                    input_ids, segment_ids, input_mask)
+            for i, example_index in enumerate(example_indices):
+                start_logits = batch_start_logits[i].detach().cpu().tolist()
+                end_logits = batch_end_logits[i].detach().cpu().tolist()
+                eval_feature = eval_features[example_index.item()]
+                unique_id = int(eval_feature.unique_id)
+                all_results.append(RawResult(unique_id=unique_id,
+                                             start_logits=start_logits,
+                                             end_logits=end_logits))
+        if self.output_dir:
+            if not os.path.exists(self.output_dir):
+                os.makedirs(self.output_dir)
+            output_prediction_file = os.path.join(self.output_dir, "predictions.json")
+            output_nbest_file = os.path.join(self.output_dir, "nbest_predictions.json")
+            output_null_log_odds_file = os.path.join(self.output_dir, "null_odds.json")
+        else:
+            output_prediction_file = None
+            output_nbest_file = None
+            output_null_log_odds_file = None
+        final_prediction, all_predictions, all_nbest_json, \
+         scores_diff_json, best_logit, n_best_predictions_dict = write_predictions(
+            eval_examples,
+            eval_features,
+            all_results,
+            self.n_best_size,
+            self.max_answer_length,
+            self.do_lower_case,
+            output_prediction_file,
+            output_nbest_file,
+            output_null_log_odds_file,
+            self.verbose_logging,
+            self.version_2_with_negative,
+            self.null_score_diff_threshold)
+        return n_best_predictions_dict
+
+
+
+
+def n_best_predictions(final_predictions_sorted, all_examples, n):
+    question_id_list = [
+        list(final_predictions_sorted.items())[i][0]
+        for i in range(len(list(final_predictions_sorted.items())))
+    ]
+    final_prediction_list = [
+        list(final_predictions_sorted.items())[i][1]['text']
+        for i in range(len(list(final_predictions_sorted.items())))
+    ]
+    title_list = [[e][0].title for q in question_id_list for e in all_examples
+                  if e.qas_id == q]
+    paragraph_list = [[e][0].paragraph for q in question_id_list
+                      for e in all_examples if e.qas_id == q]
+    final_list = [final_prediction_list, title_list, paragraph_list]
+    final_dict = collections.OrderedDict()
+    final_dict = {z[0]: list(z[1:]) for z in zip(*final_list)}
```

Comment:
Instead of returning a dictionary where the keys are the answers, I rather send a list of 3-tuples like that: `[(answer1,title1, paragraph1), (answer2, title2, paragraph2), ..., (answerN, titleN, paragraphN)]`

It's a bit weird to use answers (which are very variable strings) as keys for a dict.